lesson14.md

```diff
@@ -0,0 +1,319 @@
+<!-- .slide: id="lesson14" -->
+
+# Basic Frontend - Spring 2020
+
+Lesson 14, Thursday, 2020-04-23
+
+---
+
+### Recap - Datatypes
+
+```js
+let a1 = 42;
+let a2 = true;
+let a3 = "hello";
+let a4 = function() {};
+let a5;
+let a6 = [];
+let a7 = {};
+```
+
+---
+
+### Solution - Datatypes
+
+```js
+let a1 = 42;       // number
+let a2 = true;     // boolean
+let a3 = "hello";  // string
+let a4 = function() {};  // function
+let a5;            // undefined
+let a6 = [];       // object (array)
+let a7 = {};       // object
+```
+
+---
+
+### Null
+
+* There's one datatype that we haven't learned about yet: `null`
+* `null` is like `undefined`. It indicates the absence of a value.
+* `null` can be used to indicate an *intentional* abscence of an object value.
+* Example: `document.getElementById("NoExist");` returns `null`
+
+---
+
+### Recap - DOM
+
+* DOM is an API
+* It allows us to add/remove/change any element of a web page
+* We can make TODO lists, games and cool CSS animations
+* But how would we create a web page with "dynamic" data, for example today's weather?
+
+---
+
+### Basic building blocks
+
+We need to first learn about a few concepts:
+
+1. JSON - How to exchange data
+1. REST API - How to talk with a server
+1. `fetch` API - How to connect to a server
+
+---
+
+### Object notation
+
+* In JavaScript, we can put all keys of objects in quotes. This is purely optional:
+
+```js
+let me = {
+    "name": "John",
+    lastName: "Doe",
+    "hobbies": [ "Eat", "Sleep" ]
+};
+
+console.log(me.name); // "John"
+```
+
+---
+
+### JSON
+
+* JSON stands for **J**ava**S**cript **O**bject **N**otation
+* It is a string representation of a JavaScript object
+* All *keys* in our key value pairs must be in quotes
+* Must contain only these types:
+   * `string`, `number`, `boolean`, `array`
+   * `null`
+   * another JSON `object`
+* These types are **forbidden**:
+   * `function`, `undefined`
+
+---
+
+### JSON example
+
+```json
+{
+    "name": "Bob",
+    "instructor": true,
+    "hobbies": [ "eat", "sleep" ],
+    "age": 86,
+    "workAddress": {
+        "street": "Invalidenstraße",
+        "number": 116,
+        "city": "Berlin"
+    }
+}
+```
+
+---
+
+### object <-> JSON
+
+JavaScript offers us a an API to convert an object to JSON and vice versa:
+
+```js
+let obj = {
+    name: "Bob",
+    instructor: true
+};
+
+// turn our object to JSON:
+let json = JSON.stringify(obj);
+
+// turn our JSON back to an object:
+let jsonAsObj = JSON.parse(json);
+```
+
+---
+
+### Why JSON?
```

Comment:
I would move this slide before the JSON slide. It gives a good motivation to learn JSON. Good intro to JSON before giving the details of how JavaScript sees it.

Reply:
Done :)